### PR TITLE
[Core] Simplify system config passing from Raylet to workers

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

There are some issues we met when merging community code into Ant internal repo.

First, config values of std::string type are base64-encoded in `RayConfig::instance()`, make it unreadable in C++. In community code, the only config of std::string type is `object_spilling_config`, which is only used in Python worker. when we run `ray start`, system config items of string type are base64-encoded before passing to GCS and Raylet, but in `RayConfig::instance().initialize(...)`, these strings are not base64-decoded. It breaks some Ant internal-only config items.

To solve this issue, I think the best approach is to base64-decode inside `RayConfig::instance().initialize(...)` and base64-encode `object_spilling_config` again when passing to Python workers. The first issue solved.

Then the second issue: Java has its own copy of system config. When a Java worker loads system config via 1) or 2) as described below, all string values are base64-encoded. But in Java we don't have the type information as in C++. So there's no way to know whether we should base64-decode a value or not if Java code wants to read some of them. Fortunately, we don't read them in Java yet. Furthermore, if a Java driver wants to start a local cluster, we currently allow users to specify system config in a non-type-safe way (via `ray.raylet.config.key = value`), and there's an inconsistency between users specified items (not base64-encoded) and received items via 1) or 2) (base64-encoded). So I end up removing this functionally in Java totally. (Users are still able to specify system config in Java driver with the `ray.head-args` parameter`, which allows specifying any arguments for `ray start`, including `--system-config`.) Now issue two solved.

The final issue: I noticed that there is some tricky code in our internal repo, which adds some internal-only items in `RayConfig` to store ports and file paths, and most of them are updated dynamically. (e.g. some are updated in `main.cc` just before calling `RayConfig::instance().initialize(...)`.) I believe `RayConfig` is mostly used for creating flags and it shouldn't store any environment-related parameters. It's not designed to simplify parameter passing just because someone might think that passing a parameter from class A to class B to class C is annoying. So I think it's better to let `RayConfig::instance().initialize(...)` to handle deserialization by changing the input parameter type from `std::unordered_map<std::string, std::string>` to `std::string`. Splitting the string with "," and ";" should be done inside the "initialize" function. This way, it's not possible (or at least not easy) to change system config before `RayConfig::instance().initialize(...)` because it requires an encoded string. So I updated the code to let GCS stores the original encoded string to the backend storage, and let Raylet returns the original encoded string to workers when they are registering.

FYI, existing ways to pass system config to workers:

1) Worker pool pass system config via Java system properties in Java worker command line args. (Java only. Effective for only workers.)
2) Worker loads system config via RPC call to GCS server when starting. (Java only. Effective for both drivers and workers.)
3) Raylet passing system config to workers when workers register to Raylet. (Any worker language. Effective for both drivers and workers.)

The first two are Java only and should be deprecated. The only downside of 3) is we can't access system config in a worker process before core worker connects to Raylet.

This PR includes:
* removing Java only approaches to initialize system config.
* removing system config writing to GCS in head Raylet.
* config values of std::string type are now base64-decoded in `RayConfig`.
* `RayConfig::instance().initialize(...)` now accepts `std::string` as input.
* GCS stores system config in a "," and ";" (and also base64) encoded string to the backend storage.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
@stephanie-wang Since you introduced base64 encoding in system config, please help review. Thanks!

**Review by stephanie-wang — APPROVED** [review]
I just reviewed the C++ parts. It seems like we should just use a standard serialization format instead of having the custom method though. Maybe in the future json or yaml?

**Comment by kfstorm** [comment]
@stephanie-wang I agree with you that we should use a widely-used format for serialization. But it will make this PR more complicated.

**Comment by kfstorm** [comment]
@chaokunyang @raulchen Could you review Java part?

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc [hunk 1]
```diff
@@ -28,9 +28,7 @@ class ServiceBasedGcsClientTest : public ::testing::Test {
  public:
   ServiceBasedGcsClientTest() {
     RayConfig::instance().initialize(
-        {{"ping_gcs_rpc_server_max_retries", std::to_string(60)},
-         {"maximum_gcs_destroyed_actor_cached_count", std::to_string(10)},
-         {"maximum_gcs_dead_node_cached_count", std::to_string(10)}});
+        "ping_gcs_rpc_server_max_retries,60;maximum_gcs_destroyed_actor_cached_count,10;maximum_gcs_dead_node_cached_count,10");
```
Comment (chaokunyang) [inline src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:31]: Will colon be better than comma?

**Review by chaokunyang — APPROVED** [review]

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc (hunk 1 above)
Comment (kfstorm) [reply src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc]: I'm not going to update the format of serialized config in this PR.

**Comment by kfstorm** [comment]
> I just reviewed the C++ parts. It seems like we should just use a standard serialization format instead of having the custom method though. Maybe in the future json or yaml?

@stephanie-wang I've changed the format of serialized system config to JSON in #16669.

**Comment by mirekphd** [comment]
This is some explanation why the undocumented `ray start --system-config` works:
> (Users are still able to specify system config in Java driver with the `ray.head-args` parameter`, which allows specifying any arguments for `ray start`, including `--system-config`.)

**Pull request merged**